src/main/java/duke/command/TaskList.java

```diff
@@ -0,0 +1,250 @@
+     * @return new task list with task unmarked.
+     * @throw InvalidIndexException if array at specific index is null
+     */
+    public String unmark(int num1) throws IOException {
+        try {
+            try {
+                String original = arr[num1];
+                Parser.updateLastTaskList(new TaskList(Arrays.copyOf(arr, 100)));
+                Task newTask = new Task(String.valueOf(original.charAt(1)),
+                        original.substring(7), false);
+                arr[num1] = newTask.toString();
+                Parser.updateLastCommandDetail(arr[num1]);
+                return ("OK, I've marked this task as not done yet:" + "\n" + arr[num1]);
+            } catch (IndexOutOfBoundsException e) {
+                throw new InvalidIndexException();
+            }
+        } catch (InvalidIndexException e) {
+            return (e.getMessage());
+        }
+    }
+
+    /**
+     * Checks if element at given index of array is valid or not.
+     *
+     * @param index at which content need to be checked.
+     * @return boolean for content validity.
+     * @throw InvalidIndexException if array at specific index is null
+     */
+    public boolean checkValidIndex(int index) {
+        return (arr[index] != null);
+    }
+
+    /**
+     * Returns new task list.
+     * Deletes task at given index.
+     *
+     * @param num1 index at which task need to be deleted.
+     * @return new task list with task deleted.
+     * @throw InvalidIndexException if array at specific index is null if array at specific index is null
+     */
+    public String delete(int num1) throws IOException {
+        try {
+            try {
+                if (num1 < this.getValidLen()) {
+                    String original = arr[num1];
+                    int trace = num1;
+                    String[] originalList = new String[100];
+                    for (int k = 0; k < 100; k++) {
+                        originalList[k] = arr[k];
+                    }
+                    Parser.updateLastTaskList(new TaskList(originalList));
+                    Parser.updateLastCommandDetail(original);
+                    arr[trace] = arr[trace + 1];
+                    trace++;
+
+                    while ((trace >= 1) && (originalList[trace - 1] != null)) {
+                        arr[trace] = originalList[trace + 1];
+                        trace++;
+                    }
+
+                    return ("Noted. I've removed this task:" + "\n" + original + "\n"
+                            + String.format("Now you have %d "
+                            + "tasks in the list", this.getValidLen()));
+                }
+            } catch (IndexOutOfBoundsException e) {
+                throw new InvalidIndexException();
+            }
+        } catch (InvalidIndexException e) {
+            return (e.getMessage());
+        }
+        return new InvalidIndexException().getMessage();
```

Comment:
good catching of exception cases! 
but instead of nested `try-catch` blocks, could you try to use `try-catch-catch` instead?
`try (...) { ... } catch (...) { ... } catch (...) { ... }`